R/deployments.R

```diff
@@ -121,76 +113,55 @@ deployments <- function(appPath = ".",
                              server,
-                             hostUrl,
-                             appId,
-                             bundleId,
-                             url,
-                             when,
+                             hostUrl = NULL,
+                             appId = NULL,
+                             bundleId = NULL,
+                             url = NULL,
+                             when = as.numeric(Sys.time()),
                              lastSyncTime = as.numeric(Sys.time()),
                              metadata = list()) {
 
-  # find the username if not already supplied (may differ from account nickname)
-  if (is.null(username) && length(account) > 0) {
-    # default to empty
-    username <- ""
-    userinfo <- NULL
-    try({ userInfo <- accountInfo(account, server) }, silent = TRUE)
-    if (!is.null(userinfo$username))
-      username <- userinfo$username
-  }
-
-  # find host information
-  if (is.null(hostUrl) && length(server) > 0) {
-    hostUrl <- ""
-    serverinfo <- NULL
-    try({ serverinfo <- serverInfo(server) }, silent = TRUE)
-    if (!is.null(serverinfo$url))
-      hostUrl <- serverinfo$url
-  }
-
-  # compose the standard set of fields and append any requested
-  as.data.frame(c(
-      list(name = name,
-           title = if (is.null(title)) "" else title,
-           username = username,
-           account = account,
-           server = server,
-           hostUrl = hostUrl,
-           appId = appId,
-           bundleId = if (is.null(bundleId)) "" else bundleId,
-           url = url,
-           when = when,
-           lastSyncTime = lastSyncTime),
-      metadata),
-    stringsAsFactors = FALSE)
-}
-
-
-deploymentHistoryDir <- function() {
-  rsconnectConfigDir("deployments")
+  standard <- list(
+    name = name,
+    title = title %||% "",
+    username = username,
+    account = account,
+    server = server,
+    hostUrl = hostUrl %||% "",
+    appId = appId %||% "",
+    bundleId = bundleId %||% "",
+    url = url %||% "",
+    when = when,
+    lastSyncTime = lastSyncTime
+  )
+  c(standard, metadata)
 }
 
+# Workbench uses to show a list of recently deployed content on user dashboard
 addToDeploymentHistory <- function(appPath, deploymentRecord) {
-
-  # path to deployments files
-  history <- file.path(deploymentHistoryDir(), "history.dcf")
-  newHistory <- file.path(deploymentHistoryDir(), "history.new.dcf")
-
   # add the appPath to the deploymentRecord
   deploymentRecord$appPath <- appPath
 
   # write new history file
-  write.dcf(deploymentRecord, newHistory, width = 4096)
-  cat("\n", file = newHistory, append = TRUE)
+  newHistory <- deploymentHistoryPath(new = TRUE)
+  writeDeploymentRecord(deploymentRecord, newHistory)
 
+  history <- deploymentHistoryPath()
   # append existing history to new history
-  if (file.exists(history))
+  if (file.exists(history)) {
+    cat("\n", file = newHistory, append = TRUE)
```

Comment:
Why did the addition of a newline shift from its current location? Are we leaving a newly created DCF file without a trailing newline?

Expand the tests to confirm the behavior when creating the file and when appending to the file.

Reply:
This should be the newline that separates records in the dcf file; you're right that I should test the behaviour.

Reply:
Thanks. I'd like to see the actual file contents, but it feels as if we have lost a trailing newline on newly created files.

My theory:

New: `second-record\nfirst-record`
Old: `second-record\nfirst-record\n`

Reply:
I'm pretty sure `write.dcf()` adds that already, but I'll definitely double check.